_test/test_instrument_classes/test_IX_moderator.m

```diff
@@ -0,0 +1,202 @@
+        function self = test_IX_moderator (name)
+            self@TestCaseWithSave(name);
+            
+            % Make some moderators
+            self.mik = IX_moderator(15,30,'ikcarp',[5,25,0.13]);
+            self.mikp = IX_moderator(12,23,'ikcarp_param',[0.05,25,200],'-energy',120);
+            tri = pdf_table ([0,10,20], [0,1,0]);  % triangular distribution
+            self.mtable = IX_moderator(115,300,'table',tri);
+            self.mdelta = IX_moderator(125,10,'delta_function',[]);
+
+            self.save()
+        end
+        
+        %--------------------------------------------------------------------------
+        function test_pulse_shape_default (self)
+            mdefault = IX_moderator();
+            mdelta_zero_distAndAngle = IX_moderator(0,0,'delta_function',[]);
+            assertEqual (mdefault, mdelta_zero_distAndAngle);
+        end
+        
+        %--------------------------------------------------------------------------
+        function test_pulse_shape_mik (self)
+            t = -10:0.01:200;
+            [y, t] = pulse_shape (self.mik, t); w_mik = IX_dataset_1d (t, y);
+            assertEqualWithSave (self, w_mik);
+        end
+        
+        %--------------------------------------------------------------------------
+        function test_pulse_shape_mikp (self)
+            t = -10:0.01:200;
+            [y, t] = pulse_shape (self.mikp, t); w_mikp = IX_dataset_1d (t, y);
+            assertEqualWithSave (self, w_mikp);
+        end
+        
+        %--------------------------------------------------------------------------
+        function test_pulse_shape_mtable (self)
+            t = -10:0.01:200;
+            [y, t] = pulse_shape (self.mtable, t); w_mtable = IX_dataset_1d (t, y);
+            assertEqualWithSave (self, w_mtable);
+        end
+        
+        %--------------------------------------------------------------------------
+        function test_pulse_shape_mdelta (self)
+            t = -10:0.01:200;
+            [y, t] = pulse_shape (self.mdelta, t); w_mdelta = IX_dataset_1d (t, y);
+            assertEqualWithSave (self, w_mdelta);
+        end
+        
+        %--------------------------------------------------------------------------
+        function test_pdf_mik (self)
+            npnt = 1e7;
+            
+            % Pulse shape
+            tbin = 0:0.1:50;
+            t = (tbin(2:end) + tbin(1:end-1)) / 2;
+            y = pulse_shape (self.mik, t); w = IX_dataset_1d (t, y);
+            area = integrate(w);
+            w = w/area.val;
+            
+            % Random sampling (deterministic; return rng state afterwards)
+            S = rng(); rng(0);
+            wsamp = vals2distr (self.mik.rand(1,npnt), tbin, 'norm', 'poisson');
+            rng(S)
+            
+            % Check random sampling and 
+            [ok,mess,wdiff,chisqr] = IX_dataset_1d_same (wsamp,w,3,'rebin','chi'); 
```

Comment:
Can't find this funcion IX_dataset_1d_same and wondering if it should be 1d_samp given the context

Reply:
This function is in herbert_core/utilities/general. It is a function that has existed for a long time, as has this test. In an ideal world it would be made a method of IX_dataset, but I propose to do such house-keeping when I check in my IX_dataset refactoring.